### PR TITLE
feat: Add metric extract processor

### PR DESCRIPTION
**Pull request by jmwilliams89**
### Proposed Change
- Added the metric extract processor. See README for implementation details.
- Refactored expression and extraction logic as an internal library

##### Checklist
- [x] Changes are tested
- [ ] CI has passed

### CODE REVIEW

File: .github/dependabot.yml
Comment (cpheps) [inline .github/dependabot.yml:100]: I think we need an entry for the `processor/metricextractprocessor` module also.

File: processor/metricextractprocessor/processor.go
Comment (cpheps) [inline processor/metricextractprocessor/processor.go:33]: nit: Can we change the name to `metricExtractProcessor` or something more specific. The latest version of otel introduced the following packages `receiver`, `exporter`, `extension`. Which caused collision in some of our code where we had those same names for structs. I have a feeling at some point there will be a `processor` package as well.

**Review by cpheps — APPROVED** [review]

File: internal/expr/record.go
Comment (StefanKurek) [inline internal/expr/record.go:80]: this makes me a little sad that it is so similar to the other function, but i also don't know if i have any good suggestions on how to refactor it that are worth it

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: internal/expr/expression_test.go
Comment (StefanKurek) [inline internal/expr/expression_test.go:24]: can get rid of this if it's true that the method under test isn't needed

File: internal/expr/map.go
Comment (StefanKurek) [inline internal/expr/map.go:32]: should errors be treated the same as non matching expressions? should we have entries for non matching expressions in the results (i haven't actually looked at the code yet where this is used...more of a general question only about this function)?

File: internal/expr/expression.go
Comment (jmwilliams89) [reply internal/expr/expression.go:33]: It's used in both the logcount and metricextract processors to determine if a log record should be evaluated.

Comment (StefanKurek) [reply internal/expr/expression.go:33]: @jmwilliams89 that's my sloppy reviewing and not checking the full code 😄

File: internal/expr/map.go
Comment (jmwilliams89) [reply internal/expr/map.go:32]: There's a distinction between matching records and extracting attributes. Even if a record is matched, it doesn't mean that all of the attributes the user defines will exist on that log record. In this case, if the user defines the location of one of their attributes as `body.embedded.field` and the `embedded` field doesn't exist, the evaluate expression will return an error. This shouldn't cause the rest of the attribute extraction to fail. Instead, it just means that attribute doesn't exist in the way the user defined and should be omitted.

File: processor/metricextractprocessor/config.go
Comment (cpheps) [reply processor/metricextractprocessor/config.go:33]: Looks like it's only for unused function parameters

File: processor/metricextractprocessor/README.md
Comment (jmwilliams89) [reply processor/metricextractprocessor/README.md:84]: This is a mistake on my part. I started with the `logcount` processor README as a starting point, since the two processors share so much in common. This sentence was missed and should be removed.

Comment (StefanKurek) [reply processor/metricextractprocessor/README.md:84]: @jmwilliams89 so what is the intended behavior if there are multiple types of matches for an attribute across the logs?

File: internal/expr/map.go
Comment (StefanKurek) [reply internal/expr/map.go:32]: @jmwilliams89 in case it wasn't clear, i wasn't suggesting that no matches here should error. i just noticed that errors and no match were treated the same in that they weren't added to the results and was wondering if it was worth having a distinction (even if both should not cause a problem). sounds like no so i'll resolve

File: internal/expr/record.go
Comment (jmwilliams89) [reply internal/expr/record.go:80]: I'm personally okay with this level of naming. We could be more verbose and call it `ConvertToResourceGroupSlice`, but I'm not losing any sleep over it. If people feel strongly enough about it, I have no problem changing it.

File: processor/metricextractprocessor/README.md
Comment (StefanKurek) [reply processor/metricextractprocessor/README.md]: @jmwilliams89 i guess that makes sense and is fair...i still think it could be helpful as a reminder that there will be multiple metrics made as you had it, but i won't block on it and leave it up to you

File: processor/metricextractprocessor/factory.go
Comment (StefanKurek) [inline processor/metricextractprocessor/factory.go:50]: @jmwilliams89 minor and non blocking question - would these make more sense in the validation of the config?

File: processor/metricextractprocessor/processor.go
Comment (StefanKurek) [inline processor/metricextractprocessor/processor.go:122]: non blocking preference: would be a little nicer if we continued on the inverse of this to have a little less nesting

**Review by StefanKurek — APPROVED** [review]
A few non blocking comments/questions. But OK to merge as is

File: processor/metricextractprocessor/processor.go
Comment (jmwilliams89) [reply processor/metricextractprocessor/processor.go:122]: That's a great idea. I'll gladly make this change.

File: processor/metricextractprocessor/factory.go
Comment (jmwilliams89) [reply processor/metricextractprocessor/factory.go:50]: Sure, I think we can include that.

Comment (StefanKurek) [reply processor/metricextractprocessor/factory.go:50]: @jmwilliams89 now that you've put them in config validation, did you want to remove them from here?

Comment (StefanKurek) [reply processor/metricextractprocessor/factory.go:50]: talked about this offline

**Pull request merged**
